mql.go

```diff
@@ -0,0 +1,103 @@
+package mql
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/iancoleman/strcase"
+)
+
+// WhereClause contains a SQL where clause condition and its arguments.
+type WhereClause struct {
+	// Condition is the where clause condition
+	Condition string
+	// Args for the where clause condition
+	Args []any
+}
+
+// Parse will parse the query and use the provide database model to create a
+// where clause. Supported options: WithColumnMap, WithIgnoreFields
+func Parse(query string, model any, opt ...Option) (*WhereClause, error) {
+	const op = "mql.Parse"
+	switch {
+	case query == "":
+		return nil, fmt.Errorf("%s: missing query: %w", op, ErrInvalidParameter)
+	case isNil(model):
+		return nil, fmt.Errorf("%s: missing model: %w", op, ErrInvalidParameter)
+	}
+	p := newParser(query)
+	expr, err := p.parse()
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", op, err)
+	}
+	fValidators, err := fieldValidators(reflect.ValueOf(model), opt...)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", op, err)
+	}
+	e, err := exprToWhereClause(expr, fValidators, opt...)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", op, err)
+	}
+	return e, nil
+}
+
+// exprToWhereClause generates the where clause condition along with it's
+// required arguments. Supported options: WithColumnMap, WithConverter
+func exprToWhereClause(e expr, fValidators map[string]validator, opt ...Option) (*WhereClause, error) {
+	const op = "mql.exprToWhereClause"
+	switch {
+	case isNil(e):
+		return nil, fmt.Errorf("%s: missing expression: %w", op, ErrInvalidParameter)
+	case isNil(fValidators):
+		return nil, fmt.Errorf("%s: missing validators: %w", op, ErrInvalidParameter)
+	}
+
+	switch v := e.(type) {
+	case *comparisonExpr:
+		opts, err := getOpts(opt...)
+		if err != nil {
+			return nil, fmt.Errorf("%s: %w", op, err)
+		}
+		switch {
+		case opts.withValidateConvertColumn == v.column && !isNil(opts.withValidateConvertFn):
+			return opts.withValidateConvertFn(v.column, v.comparisonOp, v.value)
+		default:
+			columnName := strings.ToLower(v.column)
+			if n, ok := opts.withColumnMap[columnName]; ok {
+				columnName = n
+			}
+			validator, ok := fValidators[strings.ToLower(strcase.ToCamel(columnName))]
```

Comment:
Why do we need to convert to camel case and then lower case it? Is this some normalization method?

Reply:
Example: user query: `member_number = 1` with a model field name: `MemberNumber`
will not match.  Camel case will convert the user provide column name to camel
case to match the "typical" Go field names.  Make sense?

Reply:
Yes that's what I was thinking, so it is a normalization method. If we're just worried about underscores, maybe we could just strip those and avoid the third party dependency?

Reply:
done. ty!